File: src/utils/OpenAIProvider.js
/**
 * OpenAIProvider - Embedding and AI service provider
 * 
 * Capabilities:
 * - Embeddings (required): text-embedding-3-small or similar
 * - Intelligence (optional): auto-tagging, synthesis, query expansion, graph extraction
 */

export class OpenAIProvider {
  constructor(options = {}) {
    this.apiKey = options.apiKey || process.env.OPENAI_API_KEY;
    this.embeddingModel = options.embeddingModel || process.env.OPENAI_EMBEDDING_MODEL || 'text-embedding-3-small';
    this.chatModel = options.chatModel || process.env.OPENAI_CHAT_MODEL || 'gpt-4o-mini';
    this.baseUrl = options.baseUrl || 'https://api.openai.com/v1';
    this.batchSize = options.batchSize || 20;
    
    if (!this.apiKey) {
      console.warn('[OpenAIProvider] No API key provided - provider unavailable');
    }
  }

  isAvailable() {
    return !!this.apiKey;
  }

  async embed(text) {
    if (!this.apiKey) return null;
    
    const response = await fetch(`${this.baseUrl}/embeddings`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`
      },
      body: JSON.stringify({
        model: this.embeddingModel,
        input: this.cleanText(text)
      })
    });
    
    if (!response.ok) {
      const error = await response.text();
      throw new Error(`Embedding error: ${error}`);
    }
    
    const data = await response.json();
    return data.data[0].embedding;
  }

  async embedBatch(texts) {
    if (!this.apiKey || texts.length === 0) {
      return texts.map(() => null);
    }
    
    const results = [];
    
    for (let i = 0; i < texts.length; i += this.batchSize) {
      const batch = texts.slice(i, i + this.batchSize).map(t => this.cleanText(t));
      
      try {
        const response = await fetch(`${this.baseUrl}/embeddings`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${this.apiKey}`
          },
          body: JSON.stringify({
            model: this.embeddingModel,
            input: batch
          })
        });
        
        if (!response.ok) {
          console.error(`[OpenAIProvider] Batch embedding error: ${await response.text()}`);
          batch.forEach(() => results.push(null));
          continue;
        }
        
        const data = await response.json();
        data.data
          .sort((a, b) => a.index - b.index)
          .forEach(d => results.push(d.embedding));
          
      } catch (err) {
        console.error(`[OpenAIProvider] Batch error: ${err.message}`);
        batch.forEach(() => results.push(null));
      }
    }
    
    return results;
  }

  async autoTag(key, content) {
    if (!this.apiKey) return [];
    
    try {
      const response = await fetch(`${this.baseUrl}/chat/completions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.apiKey}`
        },
        body: JSON.stringify({
          model: this.chatModel,
          max_tokens: 150,
          temperature: 0,
          messages: [
            {
              role: 'system',
              content: `You are a tagging assistant. Given a document title and content, generate 8-15 relevant tags.
Return ONLY a JSON array of lowercase kebab-case tags.
Include: topic areas, technologies, concepts, and document type.
Example: ["javascript", "async-programming", "tutorial", "nodejs", "promises"]`
            },
            {
              role: 'user',
              content: `Title: ${key}\n\nContent (first 1500 chars):\n${content.substring(0, 1500)}`
            }
          ]
        })
      });
      
      if (!response.ok) {
        console.error(`[OpenAIProvider] Auto-tag error: ${await response.text()}`);
        return [];
      }
      
      const data = await response.json();
      const raw = data.choices[0].message.content.trim();
      const cleaned = raw
        .replace(/```(?:json)?/g, '')
        .replace(/```/g, '')
        .trim();
      
      return JSON.parse(cleaned);
      
    } catch (err) {
      console.error(`[OpenAIProvider] Auto-tag parse error: ${err.message}`);
      return [];
    }
  }

  async synthesize(content, task, style = '') {
    if (!this.apiKey) {
      return `## Raw Content\n\n${content.substring(0, 2000)}\n\n*(AI synthesis disabled - no API key)*`;
    }
    
    const styleGuides = {
      beginner: 'Use simple language and real-world analogies. Assume zero prior knowledge.',
      advanced: 'Be concise and technical. Focus on edge cases and implementation details.',
      visual: 'Use ASCII diagrams and step-by-step traces where helpful.',
      concise: 'Be extremely brief — max 150 words.',
      detailed: 'Be comprehensive and thorough. Include examples and edge cases.'
    };
    
    const styleGuide = styleGuides[style] || '';
    
    try {
      const response = await fetch(`${this.baseUrl}/chat/completions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.apiKey}`
        },
        body: JSON.stringify({
          model: this.chatModel,
          max_tokens: 1500,
          temperature: 0.4,
          messages: [
            {
              role: 'system',
              content: `You are a helpful AI assistant working from personal notes. ${styleGuide}
Respond in the format most appropriate for the task. Be direct and practical.
Cite your sources when possible.`
            },
            {
              role: 'user',
              content: `Task: ${task}\n\nRelevant content from notes:\n\n${content.substring(0, 7000)}`
            }
          ]
        })
      });
      
      if (!response.ok) {
        throw new Error(await response.text());
      }
      
      const data = await response.json();
      return data.choices[0].message.content.trim();
      
    } catch (err) {
      console.error(`[OpenAIProvider] Synthesis error: ${err.message}`);
      return `${content.substring(0, 1500)}\n\n*(AI synthesis failed - raw notes shown)*`;
    }
  }

  async expandQuery(query) {
    if (!this.apiKey) return [query];
    
    try {
      const response = await fetch(`${this.baseUrl}/chat/completions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.apiKey}`
        },
        body: JSON.stringify({
          model: this.chatModel,
          max_tokens: 150,
          temperature: 0,
          messages: [
            {
              role: 'system',
              content: `Given a search query, generate 10-15 related search terms.
Include: sub-topics, related concepts, prerequisites, and common aliases.
Return ONLY a JSON array of strings.`
            },
            {
              role: 'user',
              content: `Query: "${query}"`
            }
          ]
        })
      });
      
      if (!response.ok) return [query];
      
      const data = await response.json();
      const raw = data.choices[0].message.content.trim()
        .replace(/```[a-z]*/g, '')
        .replace(/```/g, '')
        .trim();
      
      const terms = JSON.parse(raw);
      return Array.isArray(terms) && terms.length > 0 ? terms : [query];
      
    } catch (err) {
      console.error(`[OpenAIProvider] Query expansion error: ${err.message}`);
      return [query];
    }
  }

  /**
   * Extract concepts and relationships for the knowledge graph
   * Returns: { concepts: string[], relationships: [{from, to, type, strength}] }
   */
  async extractGraphEntities(content) {
    if (!this.apiKey) {
      return { concepts: [], relationships: [] };
    }
    
    try {
      const response = await fetch(`${this.baseUrl}/chat/completions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.apiKey}`
        },
        body: JSON.stringify({
          model: this.chatModel,
          max_tokens: 800,
          temperature: 0,
          messages: [
            {
              role: 'system',
              content: `You are a knowledge graph extractor. Given content, extract key concepts and their relationships.
Return ONLY a JSON object in this exact format:
{
  "concepts": ["concept1", "concept2", ...],
  "relationships": [
    {"from": "concept1", "to": "concept2", "type": "related_to", "strength": 0.8}
  ]
}
Rules:
- Extract 5-15 meaningful concepts (technologies, ideas, entities, domains)
- Relationships should be meaningful (not every concept connected to every other)
- Use relationship types like: related_to, implements, part_of, prerequisite_for, uses, built_with
- Strength is 0.0 to 1.0`
            },
            {
              role: 'user',
              content: `Extract concepts and relationships from:\n\n${content.substring(0, 3000)}`
            }
          ]
        })
      });
      
      if (!response.ok) {
        console.error(`[OpenAIProvider] Graph extraction error: ${await response.text()}`);
        return { concepts: [], relationships: [] };
      }
      
      const data = await response.json();
      const raw = data.choices[0].message.content.trim()
        .replace(/```(?:json)?/g, '')
        .replace(/```/g, '')
        .trim();
      
      const parsed = JSON.parse(raw);
      return {
        concepts: Array.isArray(parsed.concepts) ? parsed.concepts : [],
        relationships: Array.isArray(parsed.relationships) ? parsed.relationships : []
      };
      
    } catch (err) {
      console.error(`[OpenAIProvider] Graph extraction parse error: ${err.message}`);
      return { concepts: [], relationships: [] };
    }
  }

  cleanText(text) {
    return text
      .substring(0, 8000)
      .replace(/\n+/g, ' ')
      .trim();
  }
}

File: src/adaptive_memory_engine/providers/openai_provider.py
"""OpenAI provider (chat + embeddings)."""
from __future__ import annotations

import json
import logging
import re
from typing import Iterable

import httpx

from .base import IntelligentProvider

log = logging.getLogger(__name__)


class OpenAIProvider(IntelligentProvider):
    name = "openai"
    dimensions = 1536  # text-embedding-3-small

    def __init__(
        self,
        api_key: str | None,
        base_url: str = "https://api.openai.com/v1",
        embedding_model: str = "text-embedding-3-small",
        chat_model: str = "gpt-4o-mini",
        timeout: float = 60.0,
    ) -> None:
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.embedding_model = embedding_model
        self.chat_model = chat_model
        self._client = httpx.Client(timeout=timeout)

    def is_available(self) -> bool:
        return bool(self.api_key)

    def _headers(self) -> dict[str, str]:
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

    def embed(self, text: str) -> list[float]:
        text = self._clean_text(text)
        r = self._client.post(
            f"{self.base_url}/embeddings",
            headers=self._headers(),
            json={"input": text, "model": self.embedding_model},
        )
        r.raise_for_status()
        return r.json()["data"][0]["embedding"]

    def embed_batch(self, texts: Iterable[str]) -> list[list[float]]:
        cleaned = [self._clean_text(t) for t in texts]
        out: list[list[float]] = []
        batch_size = 20
        for i in range(0, len(cleaned), batch_size):
            chunk = cleaned[i : i + batch_size]
            r = self._client.post(
                f"{self.base_url}/embeddings",
                headers=self._headers(),
                json={"input": chunk, "model": self.embedding_model},
            )
            r.raise_for_status()
            data = sorted(r.json()["data"], key=lambda d: d["index"])
            out.extend(d["embedding"] for d in data)
        return out

    def auto_tag(self, key: str, content: str) -> list[str]:
        prompt = (
            "Generate 3-5 concise, lowercase tags for the following content. "
            "Return ONLY a JSON array of strings, no other text.\n\n"
            f"Key: {key}\nContent: {self._clean_text(content)[:1500]}\nTags:"
        )
        text = self._chat(prompt, json_mode=True)
        return self._parse_tag_list(text)

    def synthesize(self, content: str, task: str, style: str = "concise") -> str:
        prompt = (
            f"Task: {task}\nStyle: {style}\n\n"
            f"Content:\n{self._clean_text(content)[:6000]}\n\nResponse:"
        )
        return self._chat(prompt)

    def expand_query(self, query: str) -> list[str]:
        prompt = (
            "Given the user query, generate 3 paraphrases that mean the same thing. "
            "Return ONLY a JSON array of strings.\n\n"
            f"Query: {query}\nParaphrases:"
        )
        text = self._chat(prompt, json_mode=True)
        try:
            arr = json.loads(text)
            if isinstance(arr, list):
                return [str(x) for x in arr]
        except Exception:  # noqa: BLE001
            pass
        return [query]

    def get_config(self) -> dict:
        return {
            "type": "openai",
            "embeddingModel": self.embedding_model,
            "chatModel": self.chat_model,
            "baseUrl": self.base_url,
        }

    # ---- internals ----

    def _chat(self, prompt: str, json_mode: bool = False) -> str:
        body: dict = {
            "model": self.chat_model,
            "messages": [{"role": "user", "content": prompt}],
            "temperature": 0.2,
        }
        if json_mode:
            body["response_format"] = {"type": "json_object"}
        r = self._client.post(
            f"{self.base_url}/chat/completions",
            headers=self._headers(),
            json=body,
        )
        r.raise_for_status()
        return r.json()["choices"][0]["message"]["content"]

    def _clean_text(self, text: str) -> str:
        if not text:
            return ""
        # Match Node: collapse whitespace, cap at 8000 chars
        cleaned = re.sub(r"\s+", " ", text).strip()
        return cleaned[:8000]

    def _parse_tag_list(self, text: str) -> list[str]:
        try:
            arr = json.loads(text)
            if isinstance(arr, dict):
                arr = arr.get("tags", [])
            if isinstance(arr, list):
                return [str(t).strip().lower() for t in arr if str(t).strip()][:10]
        except Exception:  # noqa: BLE001
            log.debug("auto_tag JSON parse failed, falling back to regex")
        # Fallback: extract quoted strings
        return [m.group(1).lower() for m in re.finditer(r'"([^"]+)"', text)][:10]